### PR TITLE
[CORE] Fix convertBroadcastExchangeToColumnar for non-WholeStageCodegen fallback

### PR DESCRIPTION
**Pull request by liujiayi771**
## What changes were proposed in this pull request?

Fix https://github.com/oap-project/gluten/issues/3423
Perhaps we can consider broadening the matching condition to non-gluten plans.

## How was this patch tested?

ci

### CODE REVIEW

**Comment by liujiayi771** [comment]
cc @Yohahaha

**Comment by Yohahaha** [comment]
I prefer add a UT to guard this patch.

**Comment by liujiayi771** [comment]
> I prefer add a UT to guard this patch.

I don’t know how to construct a case that includes dpp and fits this scenario. The triggering conditions are contained in a very complex sql.

**Comment by rui-mo** [comment]
> > I prefer add a UT to guard this patch.
> 
> I don’t know how to construct a case that includes dpp and fits this scenario. The triggering conditions are contained in a very complex sql.

Maybe disabling this config `spark.sql.codegen.wholeStage` could help reproduce.

**Comment by liujiayi771** [comment]
@rui-mo 
I have constructed a simple case that can trigger this bug. It requires setting spark.sql.adaptive.enabled=false, spark.sql.codegen.wholeStage=false, and spark.gluten.sql.columnar.project=false to trigger it.
```
scala.MatchError: Project [customer_id#0L]
+- VeloxColumnarToRowExec
   +- ^(1) FilterExecTransformer ((isnotnull(grade#1) AND (grade#1 = 5)) AND isnotnull(customer_id#0L))
      +- ^(1) NativeFileNativeScan parquet testdb.customers[customer_id#0L,grade#1] Batched: true, DataFilters: [isnotnull(grade#1), (grade#1 = 5), isnotnull(customer_id#0L)], Format: Parquet, Location: InMemoryFileIndex(1 paths)[hdfs://master-1-1:9000..., PartitionFilters: [], PushedFilters: [IsNotNull(grade), EqualTo(grade,5), IsNotNull(customer_id)], ReadSchema: struct<customer_id:bigint,grade:int>
 (of class org.apache.spark.sql.execution.ProjectExec)
	at io.glutenproject.expression.ExpressionConverter$.io$glutenproject$expression$ExpressionConverter$$convertBroadcastExchangeToColumnar(ExpressionConverter.scala:449)
	at io.glutenproject.expression.ExpressionConverter$$anonfun$$nestedInanonfun$transformDynamicPruningExpr$1$1$$anonfun$1.applyOrElse(ExpressionConverter.scala:492)
	at io.glutenproject.expression.ExpressionConverter$$anonfun$$nestedInanonfun$transformDynamicPruningExpr$1$1$$anonfun$1.applyOrElse(ExpressionConverter.scala:490)
```

Spark plan:
image

**Comment by rui-mo** [comment]
@liujiayi771 Thanks! Is it possible to add a unit test for this issue? Maybe `withSQLConf` could be used to disabling the configs you mentioned.

**Comment by liujiayi771** [comment]
> @liujiayi771 Thanks! Is it possible to add a unit test for this issue? Maybe `withSQLConf` could be used to disabling the configs you mentioned.

Yes. I have added four test cases to GlutenDynamicPartitionPruningSuite. Before the fix, it was triggering the bug, but currently, the tests are passing successfully.

**Review by rui-mo — APPROVED** [review]

File: gluten-ut/spark32/src/test/scala/io/glutenproject/utils/clickhouse/ClickHouseTestSettings.scala [hunk 1]
```diff
@@ -287,6 +287,10 @@ class ClickHouseTestSettings extends BackendTestSettings {
     "Gluten - SPARK-32659: Fix the data issue when pruning DPP on non-atomic type")
   enableSuite[GlutenDynamicPartitionPruningV1SuiteAEOnDisableScan].exclude(
     "Gluten - SPARK-32659: Fix the data issue when pruning DPP on non-atomic type")
+  enableSuite[GlutenDynamicPartitionPruningV1SuiteAEOffWSCGOnDisableProject].exclude(
+    "Gluten - SPARK-32659: Fix the data issue when pruning DPP on non-atomic type")
+  enableSuite[GlutenDynamicPartitionPruningV1SuiteAEOffWSCGOffDisableProject].exclude(
+    "Gluten - SPARK-32659: Fix the data issue when pruning DPP on non-atomic type")
```
Comment (rui-mo) [inline gluten-ut/spark32/src/test/scala/io/glutenproject/utils/clickhouse/ClickHouseTestSettings.scala:293]: @liujiayi771 Is there any issue for CH backend?  cc @zzcclp

Comment (liujiayi771) [reply gluten-ut/spark32/src/test/scala/io/glutenproject/utils/clickhouse/ClickHouseTestSettings.scala:293]: I noticed that CH's spark33 did not add tests such as GlutenDynamicPartitionPruningV1SuiteAEOffDisableScan. Therefore, I didn't add these tests to CH's spark33 either.

**Pull request merged**